### DISCUSSION TITLE
CTk Theme Builder (update)

### DISCUSSION THREAD

**Discussion by avalon60** [TomSchimansky/CustomTkinter#1531] [Showcase]
After a break of several months (partly demands of work as well as waiting for CustomTkinter 5 to bed in), I have resumed work on my CTk Theme Builder app. I originally started this under CustomTkinter version 4. 

![image](https://user-images.githubusercontent.com/89534395/235300691-fff30873-b13e-4f09-9a8a-e863e55caf22.png)

With the release of version 5, I faced new challenges, since the JSON files completely changed format. However, I have just about completed the re-engineering of colour processing to work with the new JSON format. The app also allows (or did under V4) you to change the widget geometry. Now tat the colour properties are working I intend to start re-engineering the widget geometry (shape - corner radius etc) processing. **I should add that the new theme JSON formats, make it much easier to design themes!**

The app allows you to change widget properties and instantly see the effects in the preview panel.  Here is a screenshot, showing my latest theme design (NightTrain) in the preview panel on the right.

![image](https://user-images.githubusercontent.com/89534395/235302319-1cbbce4b-13e3-444a-bed2-8461c6036bd2.png)

Features include:

- A colour palette, to use as a colours placeholder (more work to be done on this). One persisted for each theme mode;
- A colour harmonics generator (analogous, complementary, split-complementary, triadic and tetradic;
- Switch to render widgets as disabled;
- Right click on colour mapping widgets for copy, shade copy (copies a slightly darker shade), paste.
- Oodles of tooltips!
- Select whether to work on Dark or Light mode colours;
- Ability to copy widget colour properties of selected widget property(s) to the non selected appearance mode.
- Ability to copy  colour palette from currently selected appearance mode to the non selected appearance mode.
- Widget property views that you can select from, offering:
   1.  A basic list, allowing _all_ widget properties to be shown and worked on, or individual widgets.
   2. Categorised list, which groups widgets or widget properties, related in various ways (borders, buttons, text etc).
- Preferences for setting the control panel theme (and set light / dark mode) plus a number of other controls;
- 

Here is a screenshot of the NightTrain with the appearance mode selected as _Light_:

![image](https://user-images.githubusercontent.com/89534395/235299901-dc6eadf4-2ca8-4c12-835a-20b51ae4403e.png)

Here is the colour harmonics dialog, and a selection from the Catogorised view:

![image](https://user-images.githubusercontent.com/89534395/235302468-42928953-f95b-4350-9946-7f09d58a7ea6.png)

This dialog generates a number of shades of each generated colour, as can be seen in the image. The shade steps can be controlled via an option in the app preferences. The idea is to generate slightly different colours, for things such as hover effects. Once you have chosen a harmonics method & keystone colour, you can have it tagged to the theme, and saved so that it's available the next time you open the the theme and also open the colour harmonics dialog for the theme. 

Preferences:

![image](https://user-images.githubusercontent.com/89534395/235305648-1b34f242-01ca-451b-b1b6-158eb7c3d50a.png)

As you can see, you can enter an author name, this gets included into your theme file, under a provenance section. Example:

![image](https://user-images.githubusercontent.com/89534395/235302619-abd6c788-7cdc-44a3-9e4c-2e1846ef0a19.png)

The above section is included at the end of the theme file.

The images show the control panel, running with my Oceanix theme, which was also designed using CTk Theme Builder.

I have a few more features planned before releasing this on GitHub.

**Comment by Akascape** [comment]
@avalon60 
This looks super cool! Please upload it to github!

**Reply by avalon60** [reply]
Thanks mate. I plan to share asap, but it still needs some work. Making good progress though.

**Comment by OneBigLotus** [comment]
I'm so happy you went back to this project, excellent work!

**Comment by gamartin23** [comment]
darn i really like it, will try it!

**Comment by Borgelin** [comment]
Amazing work, can't wait to try it out.

**Comment by avalon60** [comment]
I have now replaced the "Shade copy" in the colour tiles' context menus, with in-place Lighter and Darker shade adjustments:
![238129514-14a93ded-dc15-4a0c-b51c-0bf084720f1a](https://github.com/TomSchimansky/CustomTkinter/assets/89534395/ce1a4307-a0d0-4795-b3d8-fe8c5fcd30fb)

**Reply by Akascape** [reply]
@avalon60 
Publish it to github if the project is ready for use. Then keep on pushing updates.

**Reply by avalon60** [reply]
> @avalon60 Publish it to github if the project is ready for use. Then keep on pushing updates.

I have yet to do the installation routines. I have made a start, based on an installer used on a previous project, but there's a bit to do. I don't really want to be spending ages, explaining to people how to install and configure. Also I want to simplify the assets, and if I push it out now, there will be a lot of cleanup for me to automate (or explain), in subsequent versions. I think it will take me a few weeks now, as I have been making good progress. I may also need some guidance on how to go about publishing the project docs on GitHub.

**Reply by PaulHalbig** [reply]
First of all I want to thank you for the dedication to this project, it's gonna be a blast. However I wish you would reconsider publishing the code as a work in progress. 

You could track it on a non-main branch and offer this version without any documentation, support, compability guarantee and releases.

It would help people who want to start experimenting with your application tremendously.

Don't mean to be a backseater, just expressing my thoughts.

**Comment by avalon60** [comment]
I'm under a bit of pressure with work at present,  and I'm doing my best. I just don't want to deal with a bunch of people pushing for fixes or enhancements. Please be patient. Besides, you never get a chance to make a first impression.

**Reply by PaulHalbig** [reply]
Really seems like I misjuged the tone of my message. Sorry bout that, didn't mean to be pushy.

**Reply by avalon60** [reply]
No problem.

**Comment by OneBigLotus** [comment]
Welp, I've been waiting for this a while and I've had to do the themes by hand. Anyone have any tips for doing this in the meantime? I'm genuinely horrible at making things look good

**Reply by Akascape** [reply]
@OneBigLotus 
I have a theme builder btw, it is not that feature rich but can help: https://github.com/TomSchimansky/CustomTkinter/discussions/264

**Reply by OneBigLotus** [reply]
Is it for CTK 5?

**Reply by Akascape** [reply]
Yeah, I have to update it once again ig.

**Reply by OneBigLotus** [reply]
I made a theme with it, then converted the values to v5 by hand. Still took me a few hours to do because color scheming is hard :(

**Reply by Akascape** [reply]
@OneBigLotus I have updated the theme-maker: https://github.com/Akascape/CTkThemeMaker
It is now based on the new ctk versions. You can also quick test the theme with it.
For basic color themes, just open the json file in a text editor, replace the _fg_color, hover_color_ hex strings using `Ctrl+H` and it should do the work. For other themes, I use this theme-maker.

**Reply by avalon60** [reply]
> I made a theme with it, then converted the values to v5 by hand. Still took me a few hours to do because color scheming is hard :(

I wrote a converter app. See [#1390](https://github.com/TomSchimansky/CustomTkinter/discussions/1390).

**Reply by OneBigLotus** [reply]
all these resources! I feel like we should start a skin pack with a bunch of themes, that could be fun

**Reply by avalon60** [reply]
I am going to try my best, to get CTk Theme Builder released within the next 2 weeks. If I don't manage to push it to GitHub, I will at least upload an install-able artifact to this thread. I have just about everything I wanted to get into this first version, but I have made a lot of changes to it over the last few weeks. So I will spend the next week or so, trying to make sure that it's reasonably stable.

**Reply by OneBigLotus** [reply]
> > I made a theme with it, then converted the values to v5 by hand. Still took me a few hours to do because color scheming is hard :(
> 
> I wrote a converter app. See [#1390](https://github.com/TomSchimansky/CustomTkinter/discussions/1390).

this app does not seem to function properly. the venv folder does not seem to have been included unless I extracted it incorrectly

**Reply by avalon60** [reply]
You shouldn't need a venv folder. All modules are standard. What happens if you try:

```
python ctk_theme_migrate.py  -h

```

**Reply by OneBigLotus** [reply]
I see, I didn't put the theme to convert in the right place lol

**Reply by OneBigLotus** [reply]
Key error line 96 KeyError "color"

**Reply by avalon60** [reply]
Are you absolutely sure you are pointing it at a v4 theme?

**Reply by OneBigLotus** [reply]
I am sure

**Comment by avalon60** [comment]
I find that very odd, 'color' is a main key in a V4 theme JSON. 
So at the top of the file, you would expect to see, something like:

```
{
  "color": {
    "window_bg_color": [
      "#1B2529",
      "#1B2529"
    ],
    "button": [
      "#173145",
      "#456582"
    ],
    "button_hover": [
      "#2e4c68",
      "#22333F"
    ],
    "button_border": [
      "#080e18",
      "#577584"
```

Can you post the full theme JSON? If possible in a zip file.

**Reply by OneBigLotus** [reply]
I think I figured it out, I downloaded the sweetkind json included with the theme builder assuming it was a V4 file to test the program, supplied the old one directly from inside V4 ctk and it worked. My mistake

**Reply by OneBigLotus** [reply]
After playing with the converter, it seems to butcher light mode a little when converting themes. I don't know if it's the -s argument that's messing things up but I've tried it with and without and it's a little weird. I'm comepletely fine changing it in post though so its not a problem

**Reply by avalon60** [reply]
I just used it to do a rough conversion of the V4 themes that I developed under the original V4 version of my CTk Theme Builder app. When I got the theme builder working with V5 I then refined them. I am on staycation next week, so I intend spending some time, working out how to publish and document the CTk Theme Builder app on Github.

**Reply by avalon60** [reply]
Please see  #1659 - the theme builder is now available.
